### PR TITLE
Add additional star widgets to the Information tab

### PR DESCRIPTION
**Pull request by willemferguson**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

Four additional star widgets are added to the Information tab. While "current" is added by default, the other three are only enabled if the appropriate preference has been checked in the Preferences. The information in these widgets are connected to the underlying dive structure and export to xml, git and html is provided.

### Mentions:

@dirkhh @bstoeger @sfuchs79

### CODE REVIEW

**Review by bstoeger — COMMENTED** [review]
I only looked at the code and didn't test it. Overall, it looks very clean and I like it. I have only a few coding-style comments. What I find a bit strange is that the order of the fields varies a lot.

Note to self: The undo-commands are very verbose. We should introduce a general EditDive template, which takes a pointer-to-member. This should significantly reduce the boilerplate. However, let's wait for this PR first.

My personal opinion on the feature: I think we should either take the new widgets or not. I see no reason to hide them with a preferences flag. After all, they are already somewhat hidden in the information tab. I somewhat changed my mind and think that most of them are useful, though I will probably never use "surge" myself. Of course, there remains the slippery slope argument: will we also add fields for biodiversity, colorfulness (if that is even a word), etc.?

If we are going to change the view with a preferences setting, I'd prefer going full-in: present different views for novices, professionals, photographers, etc.

I summary, I like the code and cautiously welcome the feature.

File: core/save-html.c [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include "tag.h"
 #include "trip.h"
 #include <stdio.h>
+#include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (willemferguson) [reply core/save-html.c]: Erronious. needs removal.

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp [hunk 2]
```diff
@@ -12,6 +12,9 @@
 #include "core/display.h"
 #include "core/divelist.h"
 
+#include <stdlib.h>
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: Erronious. needs removal.

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp [hunk 3]
```diff
@@ -36,6 +39,12 @@ TabDiveInformation::TabDiveInformation(QWidget *parent) : TabBase(parent), ui(ne
 	ui->gasHeadingLabel->setStyleSheet(CSS_SET_HEADING_BLUE);
 	ui->environmentHeadingLabel->setStyleSheet(CSS_SET_HEADING_BLUE);
 	ui->groupBox_visibility->setStyleSheet(CSSSetSmallLabel);
+	ui->groupBox_current->setStyleSheet(CSSSetSmallLabel);
+	ui->groupBox_wavesize->setStyleSheet(CSSSetSmallLabel);
+	ui->groupBox_surge->setStyleSheet(CSSSetSmallLabel);
+	ui->groupBox_chill->setStyleSheet(CSSSetSmallLabel);
+	if (!prefs.extraEnvironmentalDefault) // if extraEnvironmental preference is turned off
+		showCurrentWidget(FALSE, 0);  // Show current star widget at lefthand side
```
[the remark this answers is not in the document]
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: Will do.

File: desktop-widgets/tab-widgets/TabDiveInformation.cpp [hunk 4]
```diff
@@ -160,6 +169,16 @@ void TabDiveInformation::updateSalinity()
...
 }
 
+void TabDiveInformation::showCurrentWidget(bool show, int position)
+{
+		ui->groupBox_wavesize->setVisible(show);
+		ui->groupBox_surge->setVisible(show);
+		ui->groupBox_chill->setVisible(show);
+		int layoutPosition = ui->diveInfoScrollAreaLayout->indexOf(ui->groupBox_current);
+		ui->diveInfoScrollAreaLayout->takeAt(layoutPosition);
+		ui->diveInfoScrollAreaLayout->addWidget(ui->groupBox_current,6,position,1,1);
```
[the remark this answers is not in the document]
Comment (willemferguson) [reply desktop-widgets/tab-widgets/TabDiveInformation.cpp]: Not my code but needs improvement.

File: commands/command_edit.h [hunk 5]
```diff
@@ -244,7 +281,11 @@ struct PasteState {
 	QString buddy;
 	QString suit;
 	int rating;
+	int wavesize;
```
[the remark this answers is not in the document]
Comment (willemferguson) [reply commands/command_edit.h:284]: Nominally, the idea is that the order more or less (maybe somewhat naïvely) goes from the beginning of the dive to the end of the dive. Surface waves are encountered first, chill probably towards the end of the dive. After all, surface waves are also encountered at the end of the dive. But that was my reasoning for the order of the 5 widgets. 

The lgtm-com bot points at an error in the javascript code that needs improvement.

[the remark this answers is not in the document]
Comment (willemferguson) [reply commands/command_edit.h:284]: fixed (Javascript)

**Comment by dirkhh** [comment]
> I only looked at the code and didn't test it. Overall, it looks very clean and I like it. I have only a few coding-style comments. What I find a bit strange is that the order of the fields varies a lot.

There are no rules, so it's hard to complain.

> Note to self: The undo-commands are very verbose. We should introduce a general EditDive template, which takes a pointer-to-member. This should significantly reduce the boilerplate. However, let's wait for this PR first.

Yes, they do feel verbose to me.

> My personal opinion on the feature: I think we should either take the new widgets or not. I see no reason to hide them with a preferences flag. After all, they are already somewhat hidden in the information tab. I somewhat changed my mind and think that most of them are useful, though I will probably never use "surge" myself. Of course, there remains the slippery slope argument: will we also add fields for biodiversity, colorfulness (if that is even a word), etc.?

And that's the reason for making it conditional. I'd rather stop fighting back against any new feature someone proposes - as long as that proposal comes with code :-)
Instead let's establish that things that are a bit more narrow in scope can be added if hidden behind an option.

> If we are going to change the view with a preferences setting, I'd prefer going full-in: present different views for novices, professionals, photographers, etc.

That's very, very subjective. I'm not even sure how to delineate this for the new star widgets proposed here, let alone in general.
 
> I summary, I like the code and cautiously welcome the feature.

I think that's how I feel about it as well.
I am not thrilled with the fact that the review comments are addressed in a final commit. I'd much rather have a clean series. So I think I'll merge this locally and take care of that.

**Comment by dirkhh** [comment]
merged

**Comment by bstoeger** [comment]
> Instead let's establish that things that are a bit more narrow in scope can be added if hidden behind an option.

Fair enough. In this case I'm not sure what the user gains from hiding the widgets. Or is it rather that the feature is hidden from the user so that they don't demand a new filter option? 😁

**Comment by bstoeger** [comment]
> There are no rules, so it's hard to complain.

I'd have liked the order of the fields to be consistent across the code. No big deal.

**Pull request closed without merging**